### PR TITLE
Fix usages of const EVP_MD.

### LINKED ISSUE #13837: Replace the EVP_sha1() and similar calls with explicit algorithm fetch
At least in places where we have a libctx available we should never use EVP_sha1() and others but instead fetch the algorithm explicitly.

### PR DESCRIPTION
**Pull request by slontis**
Partially fixes #13837

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

Looks like you are doing this one in your next sprint.. I started on the easy stuff here..

### CODE REVIEW

File: crypto/x509/x_all.c [hunk 1]
```diff
@@ -435,23 +435,25 @@ ASN1_OCTET_STRING *X509_digest_sig(const X509 *cert)
 int X509_CRL_digest(const X509_CRL *data, const EVP_MD *type,
                     unsigned char *md, unsigned int *len)
 {
-    if (type == EVP_sha1() && (data->flags & EXFLAG_SET) != 0
+    if (type != NULL
+            && EVP_MD_is_a(SN_sha1)
```
Comment (paulidale) [inline crypto/x509/x_all.c:439]: I suggest you try to compile this :)

**Review by paulidale — CHANGES_REQUESTED** [review]

File: crypto/ts/ts_rsp_sign.c [hunk 2]
```diff
@@ -117,20 +117,33 @@ TS_RESP_CTX *TS_RESP_CTX_new(void)
...
-
+    if (propq != NULL) {
+        ctx->propq = OPENSSL_strdup(propq);
+        if (ctx->propq == NULL) {
+            OPENSSL_free(ctx);
+            ERR_raise(ERR_LIB_TS, ERR_R_MALLOC_FAILURE);
+            return NULL;
+        }
+    }
+    ctx->libctx = libctx;
```
[the remark this answers is not in the document]
Comment (mattcaswell) [reply crypto/ts/ts_rsp_sign.c:128]: That would be hard to do since a libctx doesn't have a ref count :-)

[the remark this answers is not in the document]
Comment (slontis) [reply crypto/ts/ts_rsp_sign.c:128]: libctx is used like this everywhere..

File: crypto/x509/x_all.c (hunk 1 above)
Comment (slontis) [reply crypto/x509/x_all.c]: Yep.. This was a WIP branch that I half did.. Fixed it up now.. It was more so that @t8m didnt have to redo some of the code.

**Comment by slontis** [comment]
Rebased..
ping

File: crypto/x509/x_all.c [hunk 3]
```diff
@@ -435,23 +435,25 @@ ASN1_OCTET_STRING *X509_digest_sig(const X509 *cert)
...
             && (data->flags & EXFLAG_NO_FINGERPRINT) == 0) {
         /* Asking for SHA1; always computed in CRL d2i. */
         if (len != NULL)
             *len = sizeof(data->sha1_hash);
         memcpy(md, data->sha1_hash, sizeof(data->sha1_hash));
         return 1;
     }
-    return (ASN1_item_digest
-            (ASN1_ITEM_rptr(X509_CRL), type, (char *)data, md, len));
+    return (asn1_item_digest_ex(ASN1_ITEM_rptr(X509_CRL), type, (char *)data,
```
Comment (paulidale) [inline crypto/x509/x_all.c:448]: Drop the out parenthesis?

File: crypto/x509/x_all.c [hunk 4]
```diff
@@ -435,23 +435,25 @@ ASN1_OCTET_STRING *X509_digest_sig(const X509 *cert)
...
+                                md, len, data->libctx, data->propq));
 }
 
 int X509_REQ_digest(const X509_REQ *data, const EVP_MD *type,
                     unsigned char *md, unsigned int *len)
 {
-    return (ASN1_item_digest
-            (ASN1_ITEM_rptr(X509_REQ), type, (char *)data, md, len));
+    return (asn1_item_digest_ex(ASN1_ITEM_rptr(X509_REQ), type, (char *)data,
```
Comment (paulidale) [inline crypto/x509/x_all.c:455]: Drop the out parenthesis?

File: doc/man3/TS_RESP_CTX_new.pod [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+ TS_RESP_CTX *TS_RESP_CTX_new_ex(OSSL_LIB_CTX *libctx, const char *propq);
+ TS_RESP_CTX *TS_RESP_CTX_new(void);
+ void TS_RESP_CTX_free(TS_RESP_CTX *ctx);
+
+=head1 DESCRIPTION
+
+Creates a response context that can be used for generating responses.
+
+TS_RESP_CTX_new_ex() allocates and initializes a TS_RESP_CTX structure with a
+library context of I<libctx>, property query of <propq>.
```
Comment (paulidale) [inline doc/man3/TS_RESP_CTX_new.pod:22]: `s/,/ and a/`

File: doc/man3/TS_RESP_CTX_new.pod [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+being used for any operations requiring algorithm fetches.
+
+TS_RESP_CTX_free() frees the B<TS_RESP_CTX> object I<ctx>.
+
+=head1 RETURN VALUES
+
+If the allocation fails, TS_RESP_CTX_new_ex() and TS_RESP_CTX_new() return NULL,
+otherwise it returns a pointer to the newly allocated structure.
+
+
```
Comment (paulidale) [inline doc/man3/TS_RESP_CTX_new.pod:37]: Two blank lines?

File: include/crypto/x509.h [hunk 7]
```diff
@@ -80,6 +80,8 @@ struct X509_req_st {
 
     /* Set on live certificates for authentication purposes */
     ASN1_OCTET_STRING *distinguishing_id;
+    OSSL_LIB_CTX *libctx;
+    char *propq;
```
Comment (paulidale) [inline include/crypto/x509.h:84]: Is it worthwhile considering this as a `const char *` ?
Yeah, I know it means a cast on free which is why I'm not too fussed.

Comment (slontis) [reply include/crypto/x509.h:84]: This is like this in every other struct - so if you wanted to do that it should be another PR.. I don't think it is necessary.

Comment (paulidale) [reply include/crypto/x509.h:84]: Okay, ignore my ravings.

File: crypto/x509/x_all.c (hunk 4 above)
Comment (slontis) [reply crypto/x509/x_all.c:455]: The file is full of that pattern.. Fixed.

File: doc/man3/TS_RESP_CTX_new.pod [hunk 8]
```diff
@@ -19,7 +19,7 @@ TS_RESP_CTX_free - Timestamp response context object creation
 Creates a response context that can be used for generating responses.
 
 TS_RESP_CTX_new_ex() allocates and initializes a TS_RESP_CTX structure with a
-library context of I<libctx>, property query of <propq>.
+library context of I<libctx>, and a property query of <propq>.
```
Comment (paulidale) [inline doc/man3/TS_RESP_CTX_new.pod:22]: drop the comma?

**Review by paulidale — APPROVED** [review]
One nitty thing

File: crypto/cms/cms_sd.c [hunk 9]
```diff
@@ -372,7 +372,7 @@ CMS_SignerInfo *CMS_add1_signer(CMS_ContentInfo *cms,
             ESS_SIGNING_CERT_V2 *sc2 = NULL;
             int add_sc;
 
-            if (md == EVP_sha1() || md == NULL) {
+            if (md == NULL || EVP_MD_is_a(md, SN_sha1)) {
```
Comment (levitte) [inline crypto/cms/cms_sd.c:375]: Shouldn't EVP_MD_is_a() check if the passed method is NULL, and always return false in that case?

Comment (levitte) [reply crypto/cms/cms_sd.c:375]: This is not a blocker, it can be dealt with in another PR

File: crypto/x509/x_req.c [hunk 10]
```diff
@@ -98,3 +106,37 @@ ASN1_OCTET_STRING *X509_REQ_get0_distinguishing_id(X509_REQ *x)
 {
     return x->distinguishing_id;
 }
+
+
+/*
+ * This should only be used if the X509 object was embedded inside another
+ * asn1 object and it needs a libctx to operate.
+ * Use X509_new_ex() instead if possible.
```
Comment (t8m) [inline crypto/x509/x_req.c:114]: The comment is talking about X509 where this is really about X509_REQ here.

File: doc/man3/TS_RESP_CTX_new.pod [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+ TS_RESP_CTX *TS_RESP_CTX_new_ex(OSSL_LIB_CTX *libctx, const char *propq);
+ TS_RESP_CTX *TS_RESP_CTX_new(void);
+ void TS_RESP_CTX_free(TS_RESP_CTX *ctx);
+
+=head1 DESCRIPTION
+
+Creates a response context that can be used for generating responses.
+
+TS_RESP_CTX_new_ex() allocates and initializes a TS_RESP_CTX structure with a
+library context of I<libctx> and a property query of <propq>.
```
Comment (t8m) [inline doc/man3/TS_RESP_CTX_new.pod:22]: `I<propq>`

**Review by t8m — CHANGES_REQUESTED** [review]
Some nits found

**Comment by slontis** [comment]
Addressed NITS (in last fixup commit) and rebased..

**Review by paulidale — APPROVED** [review]

**Comment by slontis** [comment]
Rebased to fix generated file. No changes were made.

**Comment by t8m** [comment]
Merged to master as e72dbd8

**Comment by paulidale** [comment]
@slontis, do you remember what was outstanding here?  The commit says it is a partial fix.

**Comment by slontis** [comment]
There are still some calls to EVP_Digest() and EVP_DigestInit_ex()
If you do a search for EVP_sha1()
EVP_PKEY_CTX_set_dh_kdf_md() also

Note that X509_digest() is fine since it uses the libctx from the cert.

**Pull request closed without merging**